File: server/internal/game/arena_test.go
package game

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"
)

// The server embeds its own copy of the map because go:embed cannot reach
// outside the package directory. This test is what keeps that copy honest: if
// someone edits the client's map without copying it across (or vice versa),
// the two simulations would silently disagree about where the walls are.
func TestEmbeddedMapMatchesClientCopy(t *testing.T) {
	clientPath := filepath.Join("..", "..", "..", "public", "maps", DefaultMapName)
	clientCopy, err := os.ReadFile(clientPath)
	if err != nil {
		t.Fatalf("reading the client's map copy at %s: %v", clientPath, err)
	}

	if !bytes.Equal(bytes.TrimSpace(clientCopy), bytes.TrimSpace(defaultMapJSON)) {
		t.Fatalf("embedded server map maps/%s has drifted from %s — copy the client's file across so both simulations agree on the arena", DefaultMapName, clientPath)
	}
}

func TestDefaultArena_LoadsObstaclesAndSpawns(t *testing.T) {
	a := DefaultArena()

	if a.Width <= 0 || a.Height <= 0 {
		t.Fatalf("expected a sized arena, got %.1fx%.1f", a.Width, a.Height)
	}
	if len(a.Obstacles) == 0 {
		t.Fatal("expected the default map to define obstacles")
	}
	if len(a.Spawns) == 0 {
		t.Fatal("expected the default map to define spawn points")
	}

	var teamA, teamB int
	for _, s := range a.Spawns {
		if s.Team == TeamA {
			teamA++
		} else {
			teamB++
		}
	}
	if teamA == 0 || teamB == 0 {
		t.Errorf("expected spawns for both teams, got A=%d B=%d", teamA, teamB)
	}
}

func TestArena_SpawnsAreNotInsideObstacles(t *testing.T) {
	a := DefaultArena()

	for _, s := range a.Spawns {
		if a.BlocksMovementAt(s.Pos, MageRadius) {
			t.Errorf("spawn for team %v at %+v is inside an obstacle", s.Team, s.Pos)
		}
		if !a.Contains(s.Pos, MageRadius) {
			t.Errorf("spawn for team %v at %+v is outside the arena bounds", s.Team, s.Pos)
		}
	}
}

func TestArena_SpawnForFallsBackWhenMapRunsOut(t *testing.T) {
	a := DefaultArena()

	// Rooms go up to 6v6 but the map defines fewer spawns per team; the
	// fallback must still land inside the arena.
	pos := a.SpawnFor(TeamA, 99)
	if !a.Contains(pos, MageRadius) {
		t.Errorf("fallback spawn %+v is outside the arena", pos)
	}
}

func TestArena_BlocksMovementInsideAnObstacle(t *testing.T) {
	a := Arena{
		Width:  24,
		Height: 16,
		Obstacles: []Obstacle{
			{Type: ObstacleRock, Position: Vec2{X: 3, Y: 0}, Radius: 0.6, BlocksMovement: true, BlocksSight: true, BlocksProjectiles: true, TopHeight: 1.0},
		},
	}

	if !a.BlocksMovementAt(Vec2{X: 3, Y: 0}, MageRadius) {
		t.Error("expected the rock's center to block movement")
	}
	if a.BlocksMovementAt(Vec2{X: 8, Y: 0}, MageRadius) {
		t.Error("expected open ground away from the rock to be walkable")
	}
}

func TestArena_ProjectilesFlyOverLowObstacles(t *testing.T) {
	a := Arena{
		Width:  24,
		Height: 16,
		Obstacles: []Obstacle{
			{Type: ObstacleFence, Position: Vec2{X: 2, Y: 0}, IsRect: true, HalfW: 2.5, HalfH: 0.12, BlocksProjectiles: true, BlocksMovement: true, TopHeight: 0.85},
		},
	}

	if !a.BlocksProjectileAt(Vec2{X: 2, Y: 0}, 0.2, 0.5) {
		t.Error("expected a low shot to be stopped by the fence")
	}
	if a.BlocksProjectileAt(Vec2{X: 2, Y: 0}, 0.2, 2.0) {
		t.Error("expected a shot arcing above the fence to pass over it")
	}
}

func TestArena_LineOfSightBlockedBySightBlocker(t *testing.T) {
	a := Arena{
		Width:  24,
		Height: 16,
		Obstacles: []Obstacle{
			{Type: ObstacleFort, Position: Vec2{X: 0, Y: 0}, IsRect: true, HalfW: 2.5, HalfH: 0.6, BlocksSight: true, BlocksMovement: true, BlocksProjectiles: true, TopHeight: 1.3},
		},
	}

	if a.HasLineOfSight(Vec2{X: -5, Y: 0}, Vec2{X: 5, Y: 0}) {
		t.Error("expected the fort to block line of sight straight through it")
	}
	if !a.HasLineOfSight(Vec2{X: -5, Y: 6}, Vec2{X: 5, Y: 6}) {
		t.Error("expected a clear line well above the fort to have line of sight")
	}
}

func TestWorld_MageCannotWalkThroughAnObstacle(t *testing.T) {
	w := NewWorldWithArena(Arena{
		Width:  24,
		Height: 16,
		Obstacles: []Obstacle{
			{Type: ObstacleFort, Position: Vec2{X: 2, Y: 0}, IsRect: true, HalfW: 0.6, HalfH: 3, BlocksMovement: true, BlocksSight: true, BlocksProjectiles: true, TopHeight: 1.3},
		},
	})
	m := w.AddMage("p1", TeamA, ElementFire, false)
	m.Position = Vec2{X: 0, Y: 0}

	w.SetInput("p1", MageInput{Move: Vec2{X: 1}})
	stepN(w, 60, SimDt) // walk straight into the wall for a full second

	if m.Position.X > 2-0.6-MageRadius+0.01 {
		t.Errorf("expected the mage to be stopped by the wall, got X=%.3f", m.Position.X)
	}
}

func TestWorld_MageSlidesAlongAWallInsteadOfSticking(t *testing.T) {
	w := NewWorldWithArena(Arena{
		Width:  24,
		Height: 16,
		Obstacles: []Obstacle{
			{Type: ObstacleFort, Position: Vec2{X: 2, Y: 0}, IsRect: true, HalfW: 0.6, HalfH: 3, BlocksMovement: true, BlocksSight: true, BlocksProjectiles: true, TopHeight: 1.3},
		},
	})
	m := w.AddMage("p1", TeamA, ElementFire, false)
	m.Position = Vec2{X: 0, Y: 0}

	// Pushing diagonally into the wall should still make progress along Y.
	w.SetInput("p1", MageInput{Move: Vec2{X: 1, Y: 1}.Normalized()})
	stepN(w, 30, SimDt)

	if m.Position.Y <= 0.5 {
		t.Errorf("expected the mage to slide along the wall in +Y, got Y=%.3f", m.Position.Y)
	}
}

File: sim/Arena.test.ts
import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { describe, expect, it } from 'vitest';
import { createObstacle } from '../src/game/Obstacle';
import { OBSTACLE_HEIGHT } from '../src/game/config';
import type { ObstacleType as ClientObstacleType } from '../src/game/types';
import { Arena, type MapData, type Obstacle, type ObstacleType } from './Arena';
import { MAGE_RADIUS, SIM_DT } from './config';
import { TEAM_A, TEAM_B } from './entities';
import { defaultArena, DEFAULT_MAP_NAME } from './defaultMap';
import { emptyInput } from './entities';
import { Vec2 } from './Vec2';
import { World } from './World';

const MAPS_DIR = join(process.cwd(), 'public', 'maps');
const ALL_TYPES: ObstacleType[] = ['tree', 'rock', 'fort', 'fence', 'prop'];

function obstacle(over: Partial<Obstacle> & Pick<Obstacle, 'type' | 'position'>): Obstacle {
  return {
    isRect: false,
    radius: 0,
    halfW: 0,
    halfH: 0,
    blocksSight: false,
    blocksProjectiles: false,
    blocksMovement: false,
    topHeight: 1,
    ...over,
  };
}

/**
 * The guard that replaced the Go server's duplicate-map test.
 *
 * There is only one copy of each map now, so drift between *files* is
 * impossible — but the sim keeps its own obstacle footprint table (the client's
 * lives in `Obstacle.ts` as closures over `Shape`, which the sim doesn't use).
 * These assertions check that table against the client's real one, so a
 * gameplay footprint can't silently differ between practice mode and online.
 */
describe('Arena — parity with the client’s obstacle tables', () => {
  it.each(ALL_TYPES)('matches the client footprint and height for %s', (type) => {
    const parsed = Arena.parse({
      width: 40,
      height: 30,
      objects: [{ type, x: 0, y: 0 }],
    });
    const mine = parsed.obstacles[0];
    const theirs = createObstacle(0, { type: type as ClientObstacleType, x: 0, y: 0 });

    expect(mine.blocksSight).toBe(theirs.blocksSight);
    expect(mine.blocksProjectiles).toBe(theirs.blocksProjectiles);
    expect(mine.blocksMovement).toBe(theirs.blocksMovement);
    expect(mine.topHeight).toBe(OBSTACLE_HEIGHT[type as ClientObstacleType]);

    if (theirs.collision.kind === 'circle') {
      expect(mine.isRect).toBe(false);
      expect(mine.radius).toBe(theirs.collision.radius);
    } else if (theirs.collision.kind === 'rect') {
      expect(mine.isRect).toBe(true);
      expect(mine.halfW).toBe(theirs.collision.halfW);
      expect(mine.halfH).toBe(theirs.collision.halfH);
    } else {
      throw new Error(`unexpected collision shape ${theirs.collision.kind}`);
    }
  });

  it('parses every map the client ships', () => {
    const files = readdirSync(MAPS_DIR).filter((f) => f.endsWith('.json'));
    expect(files.length).toBeGreaterThan(0);

    for (const file of files) {
      const data = JSON.parse(readFileSync(join(MAPS_DIR, file), 'utf8')) as MapData;
      const arena = Arena.parse(data);
      expect(arena.width, file).toBeGreaterThan(0);
      expect(arena.height, file).toBeGreaterThan(0);
      expect(arena.obstacles.length, file).toBe(data.objects.length);
    }
  });
});

describe('Arena — default map', () => {
  it('loads obstacles and spawns for both teams', () => {
    const a = defaultArena();

    expect(DEFAULT_MAP_NAME).toBe('siege1.json');
    expect(a.width).toBeGreaterThan(0);
    expect(a.height).toBeGreaterThan(0);
    expect(a.obstacles.length).toBeGreaterThan(0);
    expect(a.spawns.some((s) => s.team === TEAM_A)).toBe(true);
    expect(a.spawns.some((s) => s.team === TEAM_B)).toBe(true);
  });

  it('gives each team a Core and two Towers (GDD §5)', () => {
    const a = defaultArena();

    for (const team of [TEAM_A, TEAM_B]) {
      const mine = a.structures.filter((s) => s.team === team);
      expect(mine.filter((s) => s.kind === 'core')).toHaveLength(1);
      expect(mine.filter((s) => s.kind === 'tower')).toHaveLength(2);
    }
  });

  it('places the two sides symmetrically, so neither starts closer', () => {
    const a = defaultArena();
    const depth = (team: number): number[] =>
      a.structures
        .filter((s) => s.team === team)
        .map((s) => Math.abs(s.pos.x))
        .sort((x, y) => x - y);

    expect(depth(TEAM_A)).toEqual(depth(TEAM_B));
  });

  it('keeps every spawn inside the arena and out of obstacles', () => {
    const a = defaultArena();
    for (const s of a.spawns) {
      expect(a.blocksMovementAt(s.pos, MAGE_RADIUS), `spawn ${JSON.stringify(s.pos)}`).toBe(false);
      expect(a.contains(s.pos, MAGE_RADIUS), `spawn ${JSON.stringify(s.pos)}`).toBe(true);
    }
  });

  // Rooms go up to 6v6 but the map defines fewer spawns per team.
  it('falls back inside the arena when the map runs out of spawns', () => {
    const a = defaultArena();
    expect(a.contains(a.spawnFor(TEAM_A, 99), MAGE_RADIUS)).toBe(true);
  });
});

describe('Arena — blocking', () => {
  it('blocks movement inside an obstacle and allows open ground', () => {
    const a = new Arena(24, 16, [
      obstacle({ type: 'rock', position: new Vec2(3, 0), radius: 0.6, blocksMovement: true }),
    ]);

    expect(a.blocksMovementAt(new Vec2(3, 0), MAGE_RADIUS)).toBe(true);
    expect(a.blocksMovementAt(new Vec2(8, 0), MAGE_RADIUS)).toBe(false);
  });

  it('lets projectiles arc over a low obstacle', () => {
    const a = new Arena(24, 16, [
      obstacle({
        type: 'fence',
        position: new Vec2(2, 0),
        isRect: true,
        halfW: 2.5,
        halfH: 0.12,
        blocksProjectiles: true,
        topHeight: 0.85,
      }),
    ]);

    expect(a.blocksProjectileAt(new Vec2(2, 0), 0.2, 0.5)).toBe(true);
    expect(a.blocksProjectileAt(new Vec2(2, 0), 0.2, 2.0)).toBe(false);
  });

  it('blocks line of sight through a sight blocker only', () => {
    const a = new Arena(24, 16, [
      obstacle({
        type: 'fort',
        position: new Vec2(0, 0),
        isRect: true,
        halfW: 2.5,
        halfH: 0.6,
        blocksSight: true,
      }),
    ]);

    expect(a.hasLineOfSight(new Vec2(-5, 0), new Vec2(5, 0))).toBe(false);
    expect(a.hasLineOfSight(new Vec2(-5, 6), new Vec2(5, 6))).toBe(true);
  });
});

describe('World — obstacle collision', () => {
  const wall = (): Arena =>
    new Arena(24, 16, [
      obstacle({
        type: 'fort',
        position: new Vec2(2, 0),
        isRect: true,
        halfW: 0.6,
        halfH: 3,
        blocksMovement: true,
        blocksSight: true,
        blocksProjectiles: true,
      }),
    ]);

  it('stops a mage walking straight into a wall', () => {
    const w = new World(wall());
    const m = w.addMage('p1', TEAM_A, 'fire', false);
    m.position = Vec2.zero;

    w.setInput('p1', { ...emptyInput(), move: new Vec2(1, 0) });
    for (let i = 0; i < 60; i++) w.step(SIM_DT);

    expect(m.position.x).toBeLessThanOrEqual(2 - 0.6 - MAGE_RADIUS + 0.01);
  });

  it('slides along a wall instead of sticking when pushed diagonally', () => {
    const w = new World(wall());
    const m = w.addMage('p1', TEAM_A, 'fire', false);
    m.position = Vec2.zero;

    w.setInput('p1', { ...emptyInput(), move: new Vec2(1, 1).normalized() });
    for (let i = 0; i < 30; i++) w.step(SIM_DT);

    expect(m.position.y).toBeGreaterThan(0.5);
  });

  /*
   * Movement never walks into a blocker, but knockback, mage separation and
   * spawning can all leave a mage overlapping one — and from in there every
   * step is blocked too, so without this it stands inside the rock for the rest
   * of the match.
   */
  it('pushes a mage that ends up inside an obstacle back out of it', () => {
    const w = new World(wall());
    const m = w.addMage('p1', TEAM_A, 'fire', false);
    m.position = new Vec2(2, 0); // dead centre of the wall

    w.step(SIM_DT);

    expect(w.blockedAt(m.position)).toBe(false);
  });

  it('lets a freed mage walk away instead of staying wedged', () => {
    const w = new World(wall());
    const m = w.addMage('p1', TEAM_A, 'fire', false);
    m.position = new Vec2(2, 0);

    w.setInput('p1', { ...emptyInput(), move: new Vec2(-1, 0) });
    for (let i = 0; i < 60; i++) w.step(SIM_DT);

    expect(m.position.x).toBeLessThan(0);
  });

  it('never separates two mages into an obstacle', () => {
    const w = new World(wall());
    const a = w.addMage('p1', TEAM_A, 'fire', false);
    const b = w.addMage('p2', TEAM_A, 'fire', false);
    // Stacked just clear of the wall: the separation push is straight at it.
    a.position = new Vec2(2 - 0.6 - MAGE_RADIUS - 0.05, 0);
    b.position = new Vec2(a.position.x - 0.05, 0);

    for (let i = 0; i < 30; i++) w.step(SIM_DT);

    expect(w.blockedAt(a.position)).toBe(false);
    expect(w.blockedAt(b.position)).toBe(false);
  });
});
